### PR TITLE
fixes #58: made work in a separate thread

### LINKED ISSUE #58: Getting locationFromAddress  severely noticeably main thread
## 🐛 Bug Report

When calling `locationFromAddress()` it noticeably freezes the main thread of the application and we can not move it to separate isolate as then it returns null. 

### Expected behavior

It should work more performance-friendly. 

### Reproduction steps

just call it with some animation or try to click on the UI and it's freezed

### Configuration
Pixel 4 

Flutter 2.2.1 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 02c026b03c (4 weeks ago) • 2021-05-27 12:24:44 -0700
Engine • revision 0fdb562ac8
Tools • Dart 2.13.1

**Version:** 1.x

geocoding: ^2.0.0

**Platform:** 
- [ ] :iphone: iOS (didn't test that)
- [X] :robot: Android

### PR DESCRIPTION
**Pull request by Mayb3Nots**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Fixes #58 by using taskQueue API

### :arrow_heading_down: What is the current behavior?
Main thread on flutter will freeze

### :new: What is the new behavior (if this is a feature change)?
It wont freeze

### :boom: Does this PR introduce a breaking change?
No

### :thinking: Checklist before submitting

- [x] I made sure all projects build.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy](https://dart.dev/tools/pub/versioning).
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I followed the style guide lines ([code style guide](https://github.com/Baseflow/flutter-geocoding/blob/master/CONTRIBUTING.md)).
- [x] I updated the relevant documentation.
- [x] I rebased onto current `master`.

### CODE REVIEW

File: geocoding/android/.settings/org.eclipse.buildship.core.prefs [hunk 1]
```diff
@@ -1,11 +1,11 @@
 arguments=
 auto.sync=false
 build.scans.enabled=false
-connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(6.3))
+connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(7.4.2))
 connection.project.dir=
 eclipse.preferences.version=1
 gradle.user.home=
-java.home=/Users/maurits/Library/Java/JavaVirtualMachines/openjdk-14.0.1/Contents/Home
+java.home=C\:/Program Files/Java/jdk-18.0.1.1
```
Comment (jogboms) [inline geocoding/android/.settings/org.eclipse.buildship.core.prefs:8]: Can you revert this change?

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 2]
```diff
@@ -1,142 +1,173 @@
...
-            case "placemarkFromCoordinates":
-                onPlacemarkFromCoordinates(call, result);
-                break;
-            default:
-                result.notImplemented();
-                break;
+  private static final String TAG = "MethodCallHandlerImpl";
+  private final Geocoding geocoding;
+  private ExecutorService executor = Executors.newSingleThreadExecutor();
+  private Handler handler = new Handler(Looper.getMainLooper());
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:32]: nit-pick: `private final`

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 3]
```diff
@@ -1,142 +1,173 @@
...
+
+  /** Forwards all incoming MethodChannel calls to the given {@code geocoding}. */
+  MethodCallHandlerImpl(Geocoding geocoding) {
+    this.geocoding = geocoding;
+  }
+
+  @Override
+  public void onMethodCall(final MethodCall call, final Result result) {
+    switch (call.method) {
+    case "locationFromAddress":
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:43]: Can you revert the indentation?

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 4]
```diff
@@ -1,142 +1,173 @@
...
-                break;
-            case "placemarkFromCoordinates":
-                onPlacemarkFromCoordinates(call, result);
-                break;
-            default:
-                result.notImplemented();
-                break;
+  private static final String TAG = "MethodCallHandlerImpl";
+  private final Geocoding geocoding;
+  private ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:31]: I know this works but i am also wondering if we should also consider using the `TaskQueue` API. https://docs.flutter.dev/development/platform-integration/platform-channels#channels-and-platform-threading

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 5]
```diff
@@ -1,142 +1,173 @@
...
+  /**
+   * Clears this instance from listening to method calls.
+   *
+   * <p>Does nothing if {@link #startListening} hasn't been called, or if we're already stopped.
+   */
+  void stopListening() {
+    if (channel == null) {
+      Log.d(TAG, "Tried to stop listening when no MethodChannel had been initialized.");
+      return;
     }
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:91]: For some reason the indentation got mixed up from 4 spaces to 2 spaces

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 6]
```diff
@@ -1,142 +1,173 @@
...
-                        "NOT_FOUND",
-                        String.format("No address information found for supplied coordinates (latitude: %f, longitude: %f).", latitude, longitude),
-                        null);
-                return;
-            }
-
-            result.success(AddressMapper.toAddressHashMapList(addresses));
-        } catch (IOException ex) {
+    try {
+      final List < Address > addresses = geocoding.placemarkFromAddress(
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:109]: `List < Address >` -> `List<Address>`

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 7]
```diff
@@ -1,142 +1,173 @@
...
+    }
+  }
+
+  private void onPlacemarkFromCoordinates(final MethodCall call, final Result result) {
+    final double latitude = call.argument("latitude");
+    final double longitude = call.argument("longitude");
+    final String languageTag = call.argument("localeIdentifier");
+
+    try {
+      final List < Address > addresses = geocoding.placemarkFromCoordinates(
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:137]: `List < Address >` -> `List<Address>`

File: geocoding/android/src/main/java/com/baseflow/geocoding/utils/AddressMapper.java [hunk 8]
```diff
@@ -26,6 +26,7 @@ private static Map<String, Object> toAddressHashMap(Address address) {
         Map<String, Object> placemark = new HashMap<>();
         final String street = AddressLineParser.getStreet(address.getAddressLine(0));
 
+        placemark.put("address", address.getAddressLine(address.getMaxAddressLineIndex()));
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/utils/AddressMapper.java:29]: This is a new inclusion. Right?

File: geocoding/example/android/.settings/org.eclipse.buildship.core.prefs [hunk 9]
```diff
@@ -1,11 +1,11 @@
 arguments=
 auto.sync=false
 build.scans.enabled=false
-connection.gradle.distribution=GRADLE_DISTRIBUTION(WRAPPER)
+connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(7.1.1))
 connection.project.dir=
 eclipse.preferences.version=1
 gradle.user.home=
-java.home=/Users/maurits/Library/Java/JavaVirtualMachines/openjdk-14.0.1/Contents/Home
+java.home=C\:/Program Files/Java/jdk-18.0.1.1
```
Comment (jogboms) [inline geocoding/example/android/.settings/org.eclipse.buildship.core.prefs:8]: Can you revert this change or don't check it in

File: geocoding_platform_interface/lib/src/models/placemark.dart [hunk 10]
```diff
@@ -34,6 +36,9 @@ class Placemark {
     this.subThoroughfare,
   });
 
+  /// The full address associated with the placemark.
+  final String? address;
+
```
Comment (jogboms) [inline geocoding_platform_interface/lib/src/models/placemark.dart:41]: Can this be placed in a different PR?

File: geocoding/example/android/.settings/org.eclipse.buildship.core.prefs (hunk 9 above)
Comment (jogboms) [reply geocoding/example/android/.settings/org.eclipse.buildship.core.prefs:8]: This did not resolve correctly 🤔 @Mayb3Nots

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 11]
```diff
@@ -1,43 +1,63 @@
...
-
 /**
  * Translates incoming Geocoding MethodCalls into well formed Java function calls for {@link
  * Geocoding}.
  */
 final class MethodCallHandlerImpl implements MethodCallHandler {
     private static final String TAG = "MethodCallHandlerImpl";
     private final Geocoding geocoding;
-    @Nullable private MethodChannel channel;
+    private ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:31]: I know this works but i am also wondering if we should also consider using the TaskQueue API. https://docs.flutter.dev/development/platform-integration/platform-channels#channels-and-platform-threading

Comment (Mayb3Nots) [reply geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:31]: Do you think we should use TaskQueue API? I dont have tons of experience with native android but I will try my best to implement it if you think it is worth changing to it

File: geocoding/example/android/.settings/org.eclipse.buildship.core.prefs (hunk 9 above)
Comment (Mayb3Nots) [reply geocoding/example/android/.settings/org.eclipse.buildship.core.prefs:8]: My bad

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java (hunk 11 above)
Comment (jogboms) [reply geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:31]: I think its worth trying out as its the recommended "flutter" way

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 12]
```diff
@@ -1,20 +1,26 @@
...
+
 import androidx.annotation.Nullable;
 
 import com.baseflow.geocoding.utils.AddressMapper;
 import com.baseflow.geocoding.utils.LocaleConverter;
 
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:16]: Unused imports.

File: geocoding_platform_interface/lib/src/models/placemark.dart (hunk 10 above)
Comment (Mayb3Nots) [reply geocoding_platform_interface/lib/src/models/placemark.dart]: Hi @jogboms I looked and there seems to be other PR that already addressed this issue  #60 . Can I know why is it not being review and merged? If there is a an issue I can create a fresh PR for it.

**Review by jogboms — COMMENTED** [review]
If you could remove the unused imports then it should be good to go. I would also test this on my end.

File: geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java [hunk 13]
```diff
@@ -1,20 +1,26 @@
 package com.baseflow.geocoding;
 
 import android.location.Address;
+import android.os.Handler;
+import android.os.Looper;
```
Comment (jogboms) [inline geocoding/android/src/main/java/com/baseflow/geocoding/MethodCallHandlerImpl.java:5]: I believe this is also unused

File: geocoding_platform_interface/lib/src/models/placemark.dart (hunk 10 above)
Comment (jogboms) [reply geocoding_platform_interface/lib/src/models/placemark.dart]: Not entirely sure why it hasn't been reviewed and merged yet. I am also just an outside contributor

**Review by jogboms — APPROVED** [review]
Great work @Mayb3Nots. Its looks and tested ok.

File: geocoding/CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,6 @@
+## 2.0.5
+
+- Fixed [#58](https://github.com/Baseflow/flutter-geocoding/issues/58) getting locationFromAddress freezes main thread.
```
Comment (jogboms) [inline geocoding/CHANGELOG.md:3]: nit-pick: missing new line

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]
LGTM modules the last remarks mentioned by @jogboms regarding the unused imports and missing line break.

**Review by mvanbeusekom — APPROVED** [review]
LGTM

**Comment by Mayb3Nots** [comment]
> 

Good to hear that!

**Comment by mvanbeusekom** [comment]
Update has been published on pub.dev (version 2.0.5)

**Comment by jogboms** [comment]
Thanks @mvanbeusekom
